volatility3/framework/plugins/linux/tracing/ftrace.py

```diff
@@ -0,0 +1,311 @@
+        Those are stored in a hash table of filters that indicates the addresses hooked.
+
+        Args:
+            ftrace_ops: The ftrace_ops struct to walk through
+
+        Returns:
+            An iterable of ftrace_func_entry structs
+        """
+
+        try:
+            current_bucket_ptr = ftrace_ops.func_hash.filter_hash.buckets.first
+        except exceptions.InvalidAddressException:
+            vollog.log(
+                constants.LOGLEVEL_VV,
+                f"ftrace_func_entry list of ftrace_ops@{ftrace_ops.vol.offset:#x} is empty/invalid. Skipping it...",
+            )
+            return []
+
+        while current_bucket_ptr.is_readable():
+            yield current_bucket_ptr.dereference().cast("ftrace_func_entry")
+            current_bucket_ptr = current_bucket_ptr.next
+
+        return None
+
+    @classmethod
+    def parse_ftrace_ops(
+        cls,
+        context: interfaces.context.ContextInterface,
+        kernel_name: str,
+        known_modules: Dict[str, List[extensions.module]],
+        ftrace_ops: interfaces.objects.ObjectInterface,
+        run_hidden_modules: bool = True,
+    ) -> Optional[Iterable[ParsedFtraceOps]]:
+        """Parse an ftrace_ops struct to highlight ftrace kernel hooking.
+        Iterates over embedded ftrace_func_entry entries, which point to hooked memory areas.
+
+        Args:
+            known_modules: A dict of known modules, used to locate callbacks origin. Typically obtained through modxview.run_modules_scanners().
+            ftrace_ops: The ftrace_ops struct to parse
+            run_hidden_modules: Whether to run the hidden_modules plugin or not. Note: it won't be run, even if specified, \
+if the "hidden_modules" key is present in known_modules.
+
+        Yields:
+            An iterable of ParsedFtraceOps dataclasses, containing a selection of useful fields (callback, hook, module) related to an ftrace_ops struct
+        """
+        kernel = context.modules[kernel_name]
+        callback = ftrace_ops.func
+        callback_symbol = module_address = module_name = None
+
+        # Try to lookup within the known modules if the callback address fits
+        module = linux_utilities_modules.Modules.module_lookup_by_address(
+            context,
+            kernel.layer_name,
+            modxview.Modxview.flatten_run_modules_results(known_modules),
+            callback,
+        )
+        # Run hidden_modules plugin if a callback origin couldn't be determined (only done once, results are re-used afterwards)
+        if (
+            module is None
+            and run_hidden_modules
+            and "hidden_modules" not in known_modules
+        ):
+            vollog.info(
+                "A callback module origin could not be determined. hidden_modules plugin will be run to detect additional modules.",
+            )
+            known_modules_addresses = set(
+                context.layers[kernel.layer_name].canonicalize(module.vol.offset)
```

Comment:
This will keep looking up the layer to get to the same function (which might even be defined on the class?).  The layer should be put in a variable much earlier on...